drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
+        // that does not affect the whole SP SoC.
+        // So, use the SP_RESET GPIO.
+        //
+
+        // Setting up to inject the measurement program into the SP
+        // has several potential failures. Use this `prep` closure
+        // and `need_undo` state to keep from indenting too much.
+        let mut prep = |undo: &mut Undo| -> Result<(), ()> {
+            self.sp_reset_enter();
+            *undo |= Undo::RESET;
+
+            // Asserting SP_RESET for >1ms here works.
+            hl::sleep_for(1);
+
+            if self.dp_write_bitflags::<Dhcsr>(Dhcsr::resume()).is_err() {
+                ringbuf_entry!(Trace::DemcrWriteError);
+                *undo |= Undo::DEBUGEN;
+                return Err(());
+            }
+            *undo |= Undo::DEBUGEN;
+
+            if self
+                .dp_write_bitflags::<Demcr>(Demcr::VC_CORERESET)
+                .is_err()
+            {
+                ringbuf_entry!(Trace::DemcrWriteError);
+                *undo |= Undo::VC_CORERESET;
+                return Err(());
+            }
+            *undo |= Undo::VC_CORERESET;
+
+            self.sp_reset_leave(true);
+            *undo &= !Undo::RESET;
+
+            // 100ms max wait. Typical wait looks to be 5ms.
+            self.halt_wait(100).map_err(|_| ())?;
+
+            // Check that RESET was caught
+            if let Ok(dfsr) = self.dp_read_bitflags::<Dfsr>() {
+                if !dfsr.is_vcatch() {
+                    ringbuf_entry!(Trace::Dfsr(dfsr));
+                    ringbuf_entry!(Trace::VcCoreResetNotCaught);
+                    return Err(());
+                }
+            } else {
+                ringbuf_entry!(Trace::DfsrReadError);
+            }
+
+            // We don't want to catch the next reset.
+            if self
+                .dp_write_bitflags::<Demcr>(Demcr::from_bits_retain(0))
+                .is_err()
+            {
+                ringbuf_entry!(Trace::DemcrWriteError);
+                return Err(());
+            }
+
+            // need_undo was set appropriately
+            if let Ok(demcr) = self.dp_read_bitflags::<Demcr>() {
+                if demcr & Demcr::VC_CORERESET != Demcr::VC_CORERESET {
+                    ringbuf_entry!(Trace::VcCoreReset(false));
+                    *undo &= !Undo::VC_CORERESET;
+                } else {
+                    ringbuf_entry!(Trace::VcCoreReset(true));
+                    return Err(());
+                }
+            } else {
+                ringbuf_entry!(Trace::DemcrReadError);
+                return Err(());
+            }
+            Ok(())
+        };
+
+        // To ensures that any cleanup is done and the SP hardware is left
+        // running properly, there can only be one return at the end.
+
+        let digest = if prep(&mut need_undo).is_ok() {
+            self.do_measure_sp()
+        } else {
+            Err(())
+        };
```

Comment:
```suggestion
        let digest = prep().and_then(|()| self.do_measure_sp());
```
(this assumes you take my suggestion above of capturing `need_undo` instead of passing it in)

Reply:
LGTM